### PR TITLE
Change iterators to start using indexing with `loc`

### LINKED ISSUE #1598: Performance for `df.items` and other iterators
Looping over the Modin dataframe is slow when you use `df.items` and other iterators. Part of this is because we are still materializing the data in the driver. This does not scale.

Since we have better indexing performance, we should take advantage of that here.

### PR DESCRIPTION
**Pull request by devin-petersohn**
* Resolves #1598

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [ ] passes `flake8 modin`
- [ ] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1598 
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/iterator.py [hunk 1]
```diff
@@ -15,21 +15,21 @@
...
             axis: axis to iterate over
             func: The function to get inner iterables from
                 each partition
         """
-        self.query_compiler = query_compiler
+        self.df = df
         self.axis = axis
         self.index_iter = (
-            iter(self.query_compiler.columns)
+            zip([slice(None)] * len(self.df.columns), iter(self.df.columns))
```
Comment (dchigarev) [inline modin/pandas/iterator.py:30]: why do we need `zip` from `None` slices and column iterator if further in code we access only last field of that tuple?
```python
if self.axis == 1:
    df = self.df[key[-1]]
```

File: modin/pandas/iterator.py [hunk 2]
```diff
@@ -41,8 +41,8 @@ def __next__(self):
...
     def next(self):
         key = next(self.index_iter)
-        if self.axis:
-            df = self.query_compiler.getitem_column_array([key]).to_pandas()
+        if self.axis == 1:
+            df = self.df[key[-1]]
         else:
-            df = self.query_compiler.getitem_row_array([key]).to_pandas()
-        return next(self.func(df))
+            df = self.df.loc[key]
```
Comment (dchigarev) [inline modin/pandas/iterator.py:47]: why not `iloc`?

File: modin/pandas/dataframe.py [hunk 3]
```diff
@@ -1240,16 +1234,28 @@ def itertuples(self, index=True, name="Pandas"):
...
+                values = s._to_pandas().to_dict()
+                if index:
+                    return tuple_type(Index=s.name, **values)
+                return tuple_type(**values)
+            else:
+                values = s._to_pandas().to_numpy()
+                if index:
+                    return tuple([s.name]) + tuple(values)
+                else:
+                    return tuple(values)
```
Comment (dchigarev) [inline modin/pandas/dataframe.py:1256]: if name is `None` pandas returns named tuple with `name="Pandas"` and fields of `df.columns`. For some reason `np.testing.assert_equal` passes ok with that arguments `np.testing.assert_equal((1,2,3), Pandas(a=1, b=2, c=3))` so the tests did not fails. Maybe another function should be used for testing tuple equality.

File: modin/pandas/test/test_dataframe.py [hunk 4]
```diff
@@ -5181,8 +5181,10 @@ def test_iterrows(self, data):
             df_equals(pandas_series, modin_series)
             assert pandas_index == modin_index
 
+    @pytest.mark.parametrize("name", [None, "NotPandas", "Pandas"])
+    @pytest.mark.parametrize("index", [True, False])
     @pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
-    def test_itertuples(self, data):
+    def test_itertuples(self, name, index, data):
```
Comment (dchigarev) [inline modin/pandas/test/test_dataframe.py:5187]: could you also add a test case where `df.columns` is `MultiIndex`? It seems that your implementation would behave differently from pandas in this case

File: modin/pandas/dataframe.py (hunk 3 above)
Comment (devin-petersohn) [reply modin/pandas/dataframe.py:1256]: Default is `name="Pandas` but `name=None` creates a regular tuple: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.itertuples.html

File: modin/pandas/iterator.py (hunk 1 above)
Comment (devin-petersohn) [reply modin/pandas/iterator.py:30]: Yes it is a good question. I have changed the code to use `iloc`.

File: modin/pandas/iterator.py (hunk 2 above)
Comment (devin-petersohn) [reply modin/pandas/iterator.py]: Changed to `iloc`

File: modin/pandas/test/test_dataframe.py (hunk 4 above)
Comment (devin-petersohn) [reply modin/pandas/test/test_dataframe.py:5279]: Added test

**Comment by devin-petersohn** [comment]
@dchigarev Please take another look, I believe I address all of your comments

**Comment by dchigarev** [comment]
@devin-petersohn Yeah, thanks for the changes, but I've asked you to add test cases where **columns** are `MultiIndex`, because in that case the following problem appears, here is reproducer for 7f9fc4a:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import pandas
    import numpy as np

    data = {
        "A": ["foo", "bar", "baz"] * 4,
        "B": ["one", "two"] * 6,
        "C": list(np.arange(6)) + list(np.arange(10, 16)),
        "D": np.arange(10, 22),
    }

    multi_columns = pandas.MultiIndex.from_tuples(
        [("a", "b"), ("a", "c"), ("a", "d"), ("b", "e")]
    )

    md_df, pd_df = (
        pd.DataFrame(data),
        pandas.DataFrame(data),
    )

    md_df.columns = multi_columns
    pd_df.columns = multi_columns

    for x in pd_df.itertuples():
        print(x)  # example of output: Pandas(Index=0, _1='foo', _2='one', _3=0, _4=10)

    for x in md_df.itertuples():  # ValueError: Type names and field names must be valid identifiers: "('a', 'b')"
        print(x)

```
Going by [documentation](https://docs.python.org/3/library/collections.html#collections.namedtuple) `namedtuple` applies only string as a field name, there is a `rename` parameter to handle invalid names, which is `True` in [pandas `itertuple` implementation](https://github.com/pandas-dev/pandas/blob/v1.0.5/pandas/core/frame.py#L1031).

I've added commit suggestion below

File: modin/pandas/dataframe.py [hunk 5]
```diff
@@ -1240,16 +1234,28 @@ def itertuples(self, index=True, name="Pandas"):
...
-            df.columns = self.columns
-            df.index = [next(index_iter)]
-            return df.itertuples(index=index, name=name)
+        from collections import namedtuple
+
+        if name is not None:
+            field_names = list(self.columns)
+            if index:
+                field_names = ["Index"] + field_names
+            tuple_type = namedtuple(name, field_names)
```
Comment (dchigarev) [inline modin/pandas/dataframe.py:1243]:
```suggestion
            tuple_type = namedtuple(name, field_names, rename=True)
```

File: modin/pandas/test/test_dataframe.py [hunk 6]
```diff
@@ -5284,16 +5286,28 @@ def test_itertuples(self, data):
...
-                for modin_row, pandas_row in zip(modin_it_custom, pandas_it_custom):
-                    np.testing.assert_equal(modin_row, pandas_row)
+        mi_index_modin = pd.MultiIndex.from_tuples(
+            [(i // 4, i // 2, i) for i in range(len(modin_df))]
+        )
+        mi_index_pandas = pandas.MultiIndex.from_tuples(
+            [(i // 4, i // 2, i) for i in range(len(pandas_df))]
+        )
+        modin_df.index = mi_index_modin
+        pandas_df.index = mi_index_pandas
```
Comment (dchigarev) [inline modin/pandas/test/test_dataframe.py:5301]:
```suggestion
        mi_index_modin = pd.MultiIndex.from_tuples(
            [(i // 4, i // 2, i) for i in range(len(modin_df.columns))]
        )
        mi_index_pandas = pandas.MultiIndex.from_tuples(
            [(i // 4, i // 2, i) for i in range(len(pandas_df.columns))]
        )
        modin_df.columns = mi_index_modin
        pandas_df.columns = mi_index_pandas
```

**Comment by devin-petersohn** [comment]
@dchigarev I think I have addressed your feedback, please take a look.

**Comment by devin-petersohn** [comment]
@dchigarev I think it is okay to merge.

**Comment by dchigarev** [comment]
> @dchigarev I think it is okay to merge.

yep, sure

**Comment by devin-petersohn** [comment]
@dchigarev Per our policy, I cannot merge my own PR, will you merge it for me? 😄 Thanks!

**Comment by dchigarev** [comment]
> @dchigarev Per our policy, I cannot merge my own PR, will you merge it for me? 😄 Thanks!

Oh, okay, I'll merge it))

**Pull request merged**